### ISSUE TITLE
RabbitMQ RabbitSubscribers not hitting right function

### ISSUE THREAD

**Issue by rexchoppers** [golevelup/nestjs#266]
Hi, we've been integrating this library in order to utilize exchanges properly for RabbitMQ in our NestJS microservice. When consuming a message from the exchange, the library seems to do a "round-robin" technique of the subscribed functions rather than hitting the correct function.

Producer (PHP Although this is not the issue as it is sending to the exchange okay)

![image](https://user-images.githubusercontent.com/52251127/118871986-0326ed80-b8e0-11eb-887d-0884e332e817.png)

RabbitMQ:

![image](https://user-images.githubusercontent.com/52251127/118872331-5bf68600-b8e0-11eb-9f24-e774fa5530b0.png)

Raw Message:

anp_1           | {
anp_1           |   fields: {
anp_1           |     consumerTag: 'amq.ctag-xUOk9kgVptPb6ykXwq97Xg',
anp_1           |     deliveryTag: 46,
anp_1           |     redelivered: false,
anp_1           |     exchange: 'orka.requestreply',
anp_1           |     routingKey: 'env.user.getNotifications'
anp_1           |   },
anp_1           |   properties: {},
anp_1           |   content: <Buffer 7b 22 6d 65 73 73 61 67 65 22 3a 22 48 65 6c 6c 6f 22 7d>
anp_1           | }

Consumer service:

```
@Injectable()
export class AppService {
  getHello(): string {
    return 'Hello World!';
  }

  @RabbitSubscribe({
    exchange: 'orka.requestreply',
    routingKey: 'env.user.getNotifications',
    queue: 'cats_queue',
  })
  public async getNotifications(msg: {}, amqpMsg: ConsumeMessage) {
    console.log("Get Notifications called");
    console.log(amqpMsg);
  }


  @RabbitSubscribe({
    exchange: 'orka.requestreply',
    routingKey: 'env.user.getOneNotification',
    queue: 'cats_queue',
  })
  public async getNotificationsFunction2(msg: {}, amqpMsg: ConsumeMessage) {
    console.log("Get Notifications 2 called");
    console.log(amqpMsg);
  }
}
```

Problem: 

As you can see from the image below, the 'Get Notifications called' should be run every time as we're using a direct exchange. However, it seems to switch between the first and second functions between every request. 

![image](https://user-images.githubusercontent.com/52251127/118873825-1aff7100-b8e2-11eb-9990-f493977b7d4a.png)

Is this config related on my end or a bug with the library itself? Any help would be much appreciated

**Comment by WonderPanda** [comment]
Hey, this question comes up frequently. The issue in your setup is reusing the `cats_queue` across multiple handlers.

In this setup all messages sent to either `env.user.getNotifications` or `env.user.getOneNotification` are all sent to the same queue (think of it like a mailbox). The behavior of Rabbit is that if there are multiple consumers listening on the same queue then messages coming in will be distributed in a round robin fashion. 

This is useful if you want to scale out a worker process across multiple machines. If you have unique workloads that need to be run based on messages received at a specific routing key they should each use their own queue.

[Tutorial 2](https://www.rabbitmq.com/tutorials/tutorial-two-javascript.html) from the Rabbit documentation covers your current setup and demonstrates the round robin behavior that you're seeing

**Comment by rexchoppers** [comment]
@WonderPanda Hi, fantastic - I apologize for this being a setup issue rather than a code issue with the library. I appreciate the response. I'll get this re-configured. Thank you once again!

**Comment by WonderPanda** [comment]
@broadstone-connorbrady No need for apologies! I appreciate you taking the time to open up such a detailed issue with the relevant screenshots and code samples. Its really cool to know that you're using the NestJS RabbitMQ integration in a multi language stack where PHP is also in the picture

**Comment by rexchoppers** [comment]
@WonderPanda We'd prefer not to add PHP into the mix but that's the joy of working with legacy systems. Thank you for your advice. I've just spotted a duplicate issue at #141 too

For anyone stumbling across this or anyone who is slightly puzzled, looking at this issue provided some answers: https://stackoverflow.com/questions/32220312/rabbitmq-amqp-best-practice-queue-topic-design-in-a-microservice-architecture

**Our solution:**
We've done some digging and found the best solution for our needs by structuring events and messages in this format

![image](https://user-images.githubusercontent.com/52251127/119036343-8ca1f280-b9a8-11eb-8b4b-0f6e62e20c9b.png)

We've then made a decorator (This is a very rough prototype by the way) which we can add to our functions

```
import { RabbitSubscribe } from "@golevelup/nestjs-rabbitmq";

export function MQMessage (service: string, cmd: string): any {
  const routingKey = service + '.cmd.' + cmd;
  const queue = service + '-' +
    cmd.replace('.', '-').toLowerCase() + '-queue';

  return RabbitSubscribe({
    exchange: 'orka.requestreply',
    routingKey: routingKey,
    queue: queue,
  });
}
```

And then now we have the ideal behavior using an exchange with the direct type. 

![image](https://user-images.githubusercontent.com/52251127/119036777-089c3a80-b9a9-11eb-916f-91039e99d203.png)

If we want to broadcast an event rather than a message to all queues (In our case, for eventual consistency and the notifications to send) Then we then can create a fanout exchange later on down the line
